Fix issue 19949: Implement support for C++ ABI tags

Geod24 opened the pull request:
```
Implements C++ ABI tags with a few more restriction than the previous PR (9995).
In particular, do not follow C++ source-level conventions when impractical
(e.g. propagation). Since this is a feature that will be seldom used by users,
I favored implementation simplicity over user-friendly semantics.
```

Previous PR by @SSoulaimane : https://github.com/dlang/dmd/pull/9995
Since the author has gone AWOL and there were some discussions on the initial PR, I had a go at implementing it. I mostly used the tests from the original PR, which were very comprehensive, but tried  to keep the special casing in semantic to a minimum.
Requires https://github.com/dlang/druntime/pull/2990 to pass tests.

Geod24 committed:
Fix issue 19949 - C++ Mangling: add support for abi-tags from the Itanium ABI

Implements C++ ABI tags with a few more restriction than the previous PR (9995).
In particular, do not follow C++ source-level conventions when impractical
(e.g. propagation). Since this is a feature that will be seldom used by users,
I favored implementation simplicity over user-friendly semantics.

Co-Authored-By: سليمان السهمي  (Suleyman Sahmi) <sahmi.soulaimane@gmail.com>

Geod24 committed:
Remove cpp_stdlib outdated comment

The mangling issues have long been resolved,
and osx32 has been dropped.

thewilsonator commented:
`restruction`? Restriction?

thewilsonator commented on src/dmd/semantic2.d:
```diff
@@ -596,3 +614,101 @@ private extern(C++) final class Semantic2Visitor : Visitor
+    // `const` (and nor is `StringExp`, by extension).
+    static int predicate(const scope Expression* e1, const scope Expression* e2) nothrow
+    {
+        scope(failure) assert(0, "An exception was thrown");
```
dmd doesn't throw, so why this?

Geod24 replied:
Might be that some of those functions are not annotated as `nothrow`, will double check and fix as needed.

Geod24 replied:
Okay I realize why I didn't do this: It's a giant rabbit hole. Same issue with `const`. You'd have to make most of DMD `nothrow` for it to work, and I just felt it was too much of a change.
Also, consider that in order for this to work, the visitors need to be `nothrow` / whatever attributes we fancy, which greatly reduce the usability of DMD as a library.

Geod24 replied:
Note that this should *never* be triggered. Indeed DMD does not `throw`. `scope (failure) assert(0);` is a cheap trick to make the function `nothrow`.

thewilsonator commented on src/dmd/cppmangle.d:
```diff
@@ -935,16 +974,30 @@ private final class CppMangleVisitor : Visitor
-                buf.writestring(d.ident.toString());
-            else
+            auto varTags = ABITagContainer.forSymbol(d);
+            if (varTags)
```
why not just
```
if (ABITagContainer.forSymbol(d))
```

Geod24 replied:
Indeed, this is a leftover of a previous iteration, will fix.

Geod24 commented:
So on OSX, the compiler is too old:
```
HOST_CXX (c++): Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix
```

From a quick Google, only LLVM 3.9 [knows about ABI tags](https://releases.llvm.org/3.9.0/tools/clang/docs/ReleaseNotes.html#id4).
 3.9 was release 2016-09-02, 3.5.0 was released 2014-01-02 and 3.5.2 was released 2015-04-02.

I tested it on my machine, it works, but there's no way to test it on the auto-tester.

Geod24 committed:
d_do_test: Run nm on C++ objects on failure

This should make our life so much simpler.

Geod24 commented:
Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

MoonlightSentinel commented:
> Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

Or manually let the test suite fetch a more recent clang if necessary?

MoonlightSentinel commented on test/runnable/cpp11.d:
```diff
@@ -61,10 +61,66 @@ void test19658()
+    }
+}
+else
+    alias ABITags = gnuAbiTag;
```
This test is only disabled on win32, hence the Azure failure

Geod24 commented:
> Okay that's not going to work. Maybe it's time to have another CI provider run the C++ interop tests.

https://github.com/dlang/dmd/pull/10964

ibuclaw commented on src/dmd/semantic2.d:
```diff
@@ -558,14 +569,18 @@ private extern(C++) final class Semantic2Visitor : Visitor
+
+                        // Handles compiler-recognized `core.attribute.gnuAbiTag`
+                        if (UserAttributeDeclaration.isGNUABITag(e))
+                            doGNUABITagSemantic(e, lastTag);
```
The condition could be moved to either Compiler or Target (probably the former is best) so other compilers can extend their own compiler recognised udas - if I understand that this is the correct place to do so.

ibuclaw replied:
Ah wait, I didn't get to the bit where it's a library defined uda.

In which case, there should be tests that ensure bad things don't happen if people supply their own (incorrect) type.

Geod24 committed:
Disable runnable_cxx on POSIX's auto-tester

This is now handled by Github actions. Don't disable Windows because this hasn't been implemented yet.

MoonlightSentinel commented on posix.mak:
```diff
@@ -19,7 +19,14 @@ ifneq (,$(findstring Darwin_64_32, $(PWD)))
+# Like test, but without runnable_cxx
+auto-tester-test:
+	$(QUIET)$(MAKE) -C src -f posix.mak unittest
+	$(QUIET)$(MAKE) -C test -f Makefile
```
```suggestion
	$(QUIET)$(MAKE) -C test -f Makefile auto-tester-test
```

Geod24 replied:
Thanks 🤦‍♂

MoonlightSentinel commented on test/compilable/cpp_abi_tag_unused.d:
```diff
@@ -0,0 +1,24 @@
+REQUIRED_ARGS: -extern-std=c++11
+TEST_OUTPUT:
+---
+---
```
An empty `TEST_OUTPUT` section isn't required anymore.

MoonlightSentinel commented on test/tools/d_do_test.d:
```diff
@@ -1550,3 +1557,26 @@ void printTestFailure(string testLogName, string output_file_temp)
+ */
+void printCppSources (in const(char)[][] compiled)
+{
+    version (Posix)
```
The windows machines have some ported Posix utilities AFAICT (e.g. bash). Might be worth to try `nm` and silently ignore failures on windows.

Geod24 commented:
Runnable_cxx also needs to be disabled on semaphore ci because it uses GCC 4.8.4

Geod24 commented:
Green at last 🎉

Geod24 commented:
Hum, this went red for Ubuntu after merge... Will fix.

Geod24 commented:
Fix being worked on here: https://github.com/dlang/dmd/pull/10974
